feat: add basic API structure

joachimdekker opened the pull request:
This pull request introduces several key changes to the `AttendanceSystem.Api` project, focusing on adding new contract classes, updating controllers to use asynchronous methods, and enhancing database configuration. The most important changes are summarized below:

### New Contract Classes:
* Added `CreateCourseContract`, `CreateSessionContract`, `CreateUserContract`, `EnrollUserContract`, and `UpdateAttendanceContract` classes to the `AttendanceSystem.Api.Contracts` namespace. These classes define the data structures for creating and managing courses, sessions, users, and attendance. [[1]](diffhunk://#diff-07790e6232661a0d390da660d1af63c8f459ceb80f6ecd79e274b407be8e87afR1-R9) [[2]](diffhunk://#diff-d2a3e50e52c930b9dee5444abe9305eb675cf1b8fa66b43715fa82287768647bR1-R7) [[3]](diffhunk://#diff-665aac409a98a24fc9c7098bb230cdae192089b495007388f9c7595908b860bdR1-R20) [[4]](diffhunk://#diff-bf050351387f6024717a2e8a36b902f11bc0bb4745723b369579924882bcc97bR1-R6) [[5]](diffhunk://#diff-0b35c131d334a891de56a140981da9533490e7290196eb5cece6afc42cf771ffR1-R12)

### Controller Updates:
* Updated `CoursesController`, `SessionsController`, and `UsersController` to use asynchronous methods for handling HTTP requests. These updates include injecting corresponding services (`CourseService`, `AttendanceService`, `UserService`) and modifying method signatures to return `Task<IActionResult>`. [[1]](diffhunk://#diff-af9aef623050e79991279cc0801473b5efa0e546ae9914a70bdcc4c307f7172fL1-R68) [[2]](diffhunk://#diff-d3a3f9daf5c2f2709902ea9df9ff3da18c8a732eaeb9d9afa144c36177fbbb8eL1-R91) [[3]](diffhunk://#diff-929743da846ca2fe94e5706ba5163498fc865c36de4bc7f0306a41e4794f850eL1-R75)

### Project and Dependency Configuration:
* Added a project reference to `AttendanceSystem.Domain.Services` in the `AttendanceSystem.Api.csproj` file to include domain service dependencies.
* Updated `Program.cs` to add MVC services and configure the database context, including ensuring the database is created or migrated based on the environment. [[1]](diffhunk://#diff-c0849eea31e1ad68388cb71ec7ccdf7d24c6eb20c7ccb5feaee0e250bb3987a7R18) [[2]](diffhunk://#diff-c0849eea31e1ad68388cb71ec7ccdf7d24c6eb20c7ccb5feaee0e250bb3987a7L28-R49)

### Database Context Enhancements:
* Modified `CoursesContext` to replace `DbSet<Teacher>` with `DbSet<User>` and added `DbSet<Department>`. Updated the `OnModelCreating` method to configure entity relationships and discriminators for `User` roles.

### Service Registration:
* Updated `Services.cs` to register `CourseService`, `AttendanceService`, and `UserService` with the dependency injection container.

joachimdekker committed:
Implement API calls using services

Implement services for API calls in `CoursesController`, `SessionsController`, and `UsersController`.

* **CoursesController**
  - Inject `CourseService` into the constructor.
  - Replace logic in each method to call the corresponding method in `CourseService`.
  - Make methods async and return `Task<IActionResult>`.

* **SessionsController**
  - Inject `AttendanceService` into the constructor.
  - Replace logic in each method to call the corresponding method in `AttendanceService`.
  - Make methods async and return `Task<IActionResult>`.

* **UsersController**
  - Inject `UserService` into the constructor.
  - Replace logic in each method to call the corresponding method in `UserService`.
  - Make methods async and return `Task<IActionResult>`.

* **Services.cs**
  - Add `CourseService`, `AttendanceService`, and `UserService` to the dependency injection container.

* **CourseService.cs**
  - Create a new class `CourseService`.
  - Add methods for `GetAllCourses`, `CreateNewCourse`, `GetCourse`, `ConfigureCourse`, `DeleteCourse`, and `EnrollUser`.
  - Use `CoursesContext` to interact with the database.

joachimdekker committed:
refactor: domain models and services for type consistency

Replaced Guid with type aliases for IDs, improving consistency and readability. Updated services and controllers to support strongly typed contracts and optimized entity relationships, ensuring clearer interface definitions and enhanced application logic.

joachimdekker committed:
refactor(api): update course and user endpoints with new contracts

Modified APIs to use string IDs instead of GUIDs for flexibility and introduced new request body contracts for session and attendance management. Enhanced several services and controllers to handle type-specific user creation and refined endpoint responses for consistency.

joachimdekker committed:
fix: update routing and models to make it work

Updated alterations, contracts and controllers to let the azure functions framework work.

joachimdekker commented on AttendanceSystem.Api/Contracts/UpdateAttendanceContract.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public required string UserId { get; set; }
+    
+    [JsonConverter(typeof(JsonStringEnumConverter<AttendanceKind>))]
```
You really are stupid, aren't you?

joachimdekker commented on AttendanceSystem.Api/Contracts/EnrollUserContract.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace AttendanceSystem.Api.Contracts;
+
+public class EnrollUserContract
+{
```
Okay, I get what you say, but it should be done differently...

joachimdekker committed:
fix: make contracts members required

nicolasantolini commented on AttendanceSystem.Domain.Model/Department.cs:
```diff
@@ -2,6 +2,5 @@
 
 public class Department
 {
-    public Guid Id { get; init; } = Guid.NewGuid();
```
the department does not have an Id ? or are we using the name as primary key?

joachimdekker replied:
Yes, I think that is best. Especially since we also don't want two departments with the same name.

nicolasantolini commented on AttendanceSystem.Domain.Model/AttendanceRecord.cs:
```diff
@@ -2,7 +2,7 @@
-    public required Session Session { get; set; }
-    public required Student Student { get; set; }
+    public required SessionId Session { get; set; }
+    public required StudentId Student { get; set; }
```
maybe for clarity better to call Session -> SessionId and the same for Student -> StudentId, so when comparing you know if you are comparing an object or an int. I say this because logically if you do object.Student I expect an object and not an int

nicolasantolini commented on AttendanceSystem.Domain.Services/CourseService.cs:
```diff
@@ -0,0 +1,61 @@
+
+        course.Name = alteration.Name ?? course.Name;
+        course.Department = alteration.Department ?? course.Department;
+
```
don't you need to do _context.Courses.UpdateAsync(course)?

joachimdekker replied:
No, EF Track changes automatically

joachimdekker replied:
No, EF Track changes automatically and will update the record when one of the properties have changed.

nicolasantolini commented on AttendanceSystem.Domain.Services/AttendanceService.cs:
```diff
@@ -7,21 +7,47 @@ namespace AttendanceSystem.Domain.Services;
+    public async Task SetAttendance(Guid sessionId, string studentId, AttendanceKind kind)
     {
         var session = await GetSession(sessionId);
         session.SetAttendance(studentId, kind);
```
also here, does it really update to the database?

joachimdekker replied:
I think I left out some logic here, because right now you cannot sign up for a session I think

nicolasantolini replied:
okay I will still approve and we can focus on that in the future now that we tracked it down

joachimdekker replied:
Good!

nicolasantolini commented on AttendanceSystem.Domain.Services/UserService.cs:
```diff
@@ -1,55 +1,95 @@
+
+        user.Name = alteration.Name ?? user.Name;
+        user.Email = alteration.Email ?? user.Email;
+
```
missing update here as well

joachimdekker committed:
refactor: update property names for clarity and consistency

Renamed `Session` to `SessionId` and `Student` to `StudentId` across related models and database configurations. This improves code readability and aligns with naming conventions in the domain model.

nicolasantolini reviewed (approved):
nice!
